### PR TITLE
Updating fastai and segmentation models

### PR DESCRIPTION
**Pull request by alexdesiqueira**
In this PR, we:
* update fastai version to 2, and refactor all necessary code.
* rewrite the deep learning segmentation model: instead of returning only the Lepidoptera, it now recognizes the ruler and labels as well.
* rewrite the classification models: instead of two models, `id_position` (returning "right way up" or "upside down") and `id_gender` (when "right way up", returning "male" or "female"), we now have `id_gender` returning three classes: "upside down", "male" or "female".

### CODE REVIEW

File: mothra/binarization.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+
+def binarization(image_rgb, weights=WEIGHTS_BIN):
+    """Extract the shape of the elements in an input image using the U-net
+    deep learning architecture.
+
+    Parameters
+    ----------
+    image_rgb : (M, N, 3) ndarray
+        Input RGB image of a lepidopteran, with ruler and tags.
```
Comment (stefanv) [inline mothra/binarization.py:81]: Missing description of `weights` argument.

File: mothra/binarization.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    ruler_bin : (M, N) ndarray
+        Binary image containing the ruler in the input image.
+    lepidop_bin : (M, N) ndarray
+        Binary image containing the lepidopteran in the input image.
+    """
+    if isinstance(weights, str):
+        weights = Path(weights)
+
+    connection.download_weights(weights)
+    # parameters here were defined when training the U-net.
```
Comment (stefanv) [inline mothra/binarization.py:96]: What does this comment mean?

File: mothra/binarization.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    -------
+    image_bin : (M, N) ndarray
+        The input binary image containing only the largest region.
+    """
+    props = regionprops(label(image_bin))
+
+    # largest_reg will receive the largest region label and its area.
+    largest_reg = [0, 0]
+    for prop in props:
+            if prop.area > largest_reg[1]:
```
Comment (stefanv) [inline mothra/binarization.py:129]:
```
max(props, key=lambda x: x.area)
```

File: mothra/binarization.py [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+@memory.cache(ignore=['axes'])
+def main(image_rgb, axes=None):
+    """Binarizes and crops the lepidopteran in image_rgb.
+
+    Parameters
+    ----------
+    image_rgb : 3D array
+        RGB image of the entire picture
+    top_ruler: integer
```
Comment (stefanv) [inline mothra/binarization.py:145]: `top_ruler` is no longer in function signature

File: mothra/identification.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
+from fastai.vision.learner import load_learner
+from pathlib import Path
+from mothra import connection
+
+
+WEIGHTS_GENDER = './models/id_gender_test-3classes.pkl'
+
+
+def _classification(image_rgb, weights):
+    """Helping function. Classifies the input image according to `weights`.
```
Comment (stefanv) [inline mothra/identification.py:10]: Helper function

File: mothra/identification.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+    Parameters
+    ----------
+    bfly_rgb : 3D array
```
Comment (stefanv) [inline mothra/identification.py:14]: Does not match signature

File: mothra/ruler_detection.py [hunk 7]
```diff
@@ -92,21 +74,20 @@ def remove_numbers(focus):
 
 
 def fourier(signal, axes=None):
-    '''Performs a fourier transform to find the frequency and t space
+    """Performs a Fourier transform to find the frequency and t space
```
Comment (stefanv) [inline mothra/ruler_detection.py:77]: This was here before, but we should probably clarify `t space`

File: pipeline.py [hunk 8]
```diff
@@ -72,6 +77,62 @@ def create_layout(n_stages, plot_level):
...
+                 'position',
+                 'gender']
+
+    with open(csv_fname, 'w') as csv_file:
+        writer = csv.writer(csv_file)
+        writer.writerow(DATA_COLS)
+    return csv_fname
+
+
+# required by fastai while predicting:
```
Comment (stefanv) [inline pipeline.py:113]: What are albumentations?

**Review by stefanv — COMMENTED** [review]
This looks good @alexdesiqueira!

Feel free to merge after addressing the (minor) comments.

File: mothra/binarization.py (hunk 2 above)
Comment (alexdesiqueira) [reply mothra/binarization.py:96]: It meant something in the past; deleting it.

File: pipeline.py (hunk 8 above)
Comment (alexdesiqueira) [reply pipeline.py:113]: It's the name of the [package](https://albumentations.ai/) embedded in fastai for augmentations.

Comment (alexdesiqueira) [reply pipeline.py:113]: Improved that comment.

File: mothra/binarization.py (hunk 3 above)
Comment (alexdesiqueira) [reply mothra/binarization.py:130]: That's great!

**Comment by alexdesiqueira** [comment]
Thank you @stefanv! All fixed, merging.

**Pull request merged**